pkg/process_manager.go

```diff
@@ -18,46 +26,335 @@ const (
 )
 
 type ProcessStatus struct {
-	// ExitStatus is only valid if `State` is `EXITED`.
-	ExitStatus int32
-	State      ProcessState
+	// ExitCode is only valid if `State` is `EXITED`.
+	ExitCode int
+	State    ProcessState
+}
+
+// multiBufferWriter is the write half of a shared buffer that allows multiple
+// readers.
+type multiBufferWriter struct {
+	bufLock sync.RWMutex
+	buf     []byte
+}
+
+func (s *multiBufferWriter) Write(p []byte) (n int, err error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	s.bufLock.Lock()
+	defer s.bufLock.Unlock()
+	s.buf = append(s.buf, p...)
+
+	return len(p), nil
+}
+
+func (s *multiBufferWriter) newReader() multiBufferReader {
+	return multiBufferReader{writer: s}
+}
+
+// multiBufferReader is the read half of a shared buffer that allows multiple
+// readers.
+type multiBufferReader struct {
+	writer *multiBufferWriter
+
+	// pos tracks how far into buf has been read already.
+	pos int
+}
+
+func (s *multiBufferReader) Read(p []byte) (n int, err error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	s.writer.bufLock.RLock()
+	defer s.writer.bufLock.RUnlock()
+	n = copy(p, s.writer.buf[s.pos:])
+
+	if n == 0 {
+		return 0, io.EOF
+	}
+	s.pos += n
+
+	return n, nil
 }
 
 type processStream struct {
-	// TODO
+	buf multiBufferReader
+
+	// caughtUp indicates whether Read has read up to the end of buf before eof
+	// became true.
+	caughtUp bool
+
+	// eof represents whether the process stream has closed.
+	eof *atomic.Bool
+
+	// waiter can be used to block until there is new data in buf from the
+	// process.
+	waiter *sync.Cond
 }
 
-func (*processStream) Read(p []byte) (n int, err error) {
-	panic("TODO")
+// copy makes a copy of a processStream with unique reading state but copied
+// buffer source and synchronization state.
+func (s *processStream) copy() *processStream {
+	return &processStream{
+		buf:    s.buf.writer.newReader(),
+		eof:    s.eof,
+		waiter: s.waiter,
+	}
+}
+
+func (s *processStream) Read(p []byte) (n int, err error) {
+	// If reads have caught up and the stream isn't closed, wait for more data.
+	if s.caughtUp && !s.eof.Load() {
+		func() {
+			s.waiter.L.Lock()
+			defer s.waiter.L.Unlock()
+			s.waiter.Wait()
+		}()
+	}
+
+	// Read historic or new data.
+	n, err = s.buf.Read(p)
```

Comment:
Your reader always returns `n==0` when it errors, so there's not really a problem here, but the general `io.Reader` contract is that it's perfectly valid for `Read` to return `n>0` and an error.

The code would be a little more robust against future changes if we handled this case.